source/lac/trilinos_epetra_vector.cc

```diff
@@ -0,0 +1,529 @@
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  namespace EpetraWrappers
+  {
+    Vector::Vector()
+      :
+      vector(new Epetra_FEVector(Epetra_Map(0,1,0,Epetra_MpiComm(MPI_COMM_SELF))))
+    {}
+
+
+
+    Vector::Vector(const Vector &V)
+      :
+      vector(new Epetra_FEVector(V.trilinos_vector()))
+    {}
+
+
+
+    Vector::Vector(const IndexSet &parallel_partitioner,
+                   const MPI_Comm &communicator)
+      :
+      vector(new Epetra_FEVector(parallel_partitioner.make_trilinos_map(communicator,false)))
+    {}
+
+
+
+    void Vector::reinit(const IndexSet &parallel_partitioner,
+                        const MPI_Comm &communicator,
+                        const bool      omit_zeroing_entries)
+    {
+      Epetra_Map input_map = parallel_partitioner.make_trilinos_map(communicator,false);
+      if (vector->Map().SameAs(input_map)==false)
+        vector.reset(new Epetra_FEVector(input_map));
+      else if (omit_zeroing_entries==false)
+        {
+          const int ierr = vector->PutScalar(0.);
+          (void) ierr;
+          Assert(ierr==0, ExcTrilinosError(ierr));
+        }
+    }
+
+
+
+    Vector &Vector::operator= (const Vector &V)
+    {
+      // Distinguish three cases:
+      //  - First case: both vectors have the same layout.
+      //  - Second case: both vectors have the same size but different layout.
+      //  - Third case: the vectors have different size.
+      if (vector->Map().SameAs(V.trilinos_vector().Map()))
+        *vector = V.trilinos_vector();
+      else
+        {
+          if (size()==V.size())
+            {
+              Epetra_Import data_exchange(vector->Map(), V.trilinos_vector().Map());
+
+              const int ierr = vector->Import(V.trilinos_vector(), data_exchange, Insert);
+              Assert(ierr==0, ExcTrilinosError(ierr));
+            }
+          else
+            vector.reset(new Epetra_FEVector(V.trilinos_vector()));
+        }
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator= (const ::dealii::LinearAlgebra::ReadWriteVector<double> &V)
+    {
+      IndexSet elements_to_write(V.get_stored_elements());
+      IndexSet locally_owned_elements(this->locally_owned_elements());
+      IndexSet off_proc_elements(elements_to_write);
+      off_proc_elements.subtract_set(locally_owned_elements);
+      IndexSet on_proc_elements(elements_to_write);
+      on_proc_elements.subtract_set(off_proc_elements);
+
+      // Write the elements that locally owned.
+      IndexSet::ElementIterator elem = on_proc_elements.begin(),
+                                elem_end = on_proc_elements.end();
+      for (; elem!=elem_end; ++elem)
+        {
+          const TrilinosWrappers::types::int_type local_index =
+            vector->Map().LID(static_cast<TrilinosWrappers::types::int_type>(*elem));
+          (*vector)[0][local_index] = V[*elem];
+        }
+
+
+      // Write the elements off-processor. Cannot use Import function of Trilinos
```

Comment:
@bangerth The current function does not what we want to do because we forgot that about ```VectorOperation::Add``` when we designed ```VectorSpaceVector``` and thus, we only have ```operation= (ReadWriteVector)``` basically we can only do ```VectorOperation::Insert```. Now, I could add something like ```operation+= (ReadWriteVector)``` or we could have something like ```import(VectorOperation::add/VectorOperation::insert)```. Then like you said, I could copy the ```ReadWriteVector``` to a ```Epetra_MultiVector``` and let Trilinos do all the work. My point is that this interface would be bad because every time we would need to rediscover the communication pattern. That's why I would like the new interface to explicitly require an object which describes the communication pattern. This is not a problem in the current interface because when you create a ghosted vector, you can basically "lock" the communication pattern since you know the two ```IndexSet``` when you call the constructor. However, this is not ideal because I still have to recompute the communication pattern for different vectors with the same layout.